File: NetFramework/Schedule/VB/MultiUserCustomForms/RibbonControls/ComboBoxHost.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Drawing.Design
Imports System.Windows.Forms
Imports System.Windows.Forms.Design
Imports System.Reflection
Imports System.Data
Imports C1.Win.C1Ribbon

<System.ComponentModel.DefaultBindingProperty("SelectedValue")> _
Public Class ComboBoxHost
    Inherits C1.Win.C1Ribbon.RibbonControlHost
    Implements IBindableComponent

    Public Sub New()
        Me.New(New System.Windows.Forms.ComboBox())
    End Sub

    Public Sub New(ByVal box)
        MyBase.New(box)
    End Sub

    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public ReadOnly Property ComboBox() As ComboBox
        Get
            Return CType(Control, ComboBox)
        End Get
    End Property

#Region "** object model"

    <Browsable(False), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Property BindingContext() As BindingContext Implements IBindableComponent.BindingContext
        Get
            Return ComboBox.BindingContext
        End Get

        Set(ByVal value As BindingContext)
            ComboBox.BindingContext = value
        End Set
    End Property

    <Category("Data")> _
    Public ReadOnly Property DataBindings() As ControlBindingsCollection Implements IBindableComponent.DataBindings
        Get
            Return ComboBox.DataBindings
        End Get
    End Property

    <Category("Data"), Browsable(False), Bindable(True), _
    DefaultValue(CStr(Nothing)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Property SelectedValue() As Object
        Get
            Return ComboBox.SelectedValue
        End Get
        Set(ByVal value As Object)
            ComboBox.SelectedValue = value
        End Set
    End Property

    <Category("Data"), Browsable(False), Bindable(True), _
    DefaultValue(CStr(Nothing)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Property SelectedItem() As Object
        Get
            Return ComboBox.SelectedItem
        End Get
        Set(ByVal value As Object)
            ComboBox.SelectedItem = value
        End Set
    End Property

    <Category("Data"), Description("Indicates the list that this control will use to get its items."), _
    AttributeProvider("System.ComponentModel.IListSource"), _
    DefaultValue(CStr(Nothing)), _
    RefreshProperties(RefreshProperties.Repaint)> _
    Public Property DataSource() As Object
        Get
            Return ComboBox.DataSource
        End Get
        Set(ByVal value As Object)
            ComboBox.DataSource = value
        End Set
    End Property

    <Category("Data"), Description("Indicates the property to display for the items in the control."), _
    Editor("System.Windows.Forms.Design.DataMemberFieldEditor, System.Design", "System.Drawing.Design.UITypeEditor"), _
    TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, System.Design"), _
    DefaultValue("")> _
    Public Property DisplayMember() As String
        Get
            Return ComboBox.DisplayMember
        End Get

        Set(ByVal value As String)
            ComboBox.DisplayMember = value
        End Set
    End Property

    <Category("Data"), Description("Indicates the property to use as the actual value for the items in the control."), _
    Editor("System.Windows.Forms.Design.DataMemberFieldEditor, System.Design", "System.Drawing.Design.UITypeEditor"), _
    TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, System.Design"), _
    DefaultValue("")> _
    Public Property ValueMember() As String
        Get
            Return ComboBox.ValueMember
        End Get

        Set(ByVal value As String)
            ComboBox.ValueMember = value
        End Set
    End Property

    <Category("Data"), Description("Gets the collection of items in this ComboBox.")> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    <Editor("System.Windows.Forms.Design.ListControlStringCollectionEditor, System.Design", GetType(UITypeEditor))> _
    Public ReadOnly Property Items() As ComboBox.ObjectCollection
        Get
            Return ComboBox.Items
        End Get
    End Property

    <Category("Behavior"), Description("Indicates whether the code or the operating system will handle drawing of elements in the list.")> _
    <DefaultValue(DrawMode.Normal)> _
    Public Property DrawMode() As DrawMode
        Get
            Return ComboBox.DrawMode
        End Get

        Set(ByVal value As DrawMode)
            ComboBox.DrawMode = value
        End Set
    End Property

    <Category("Appearance"), Description("Controls the appearance and functionality of the ComboBox.")> _
    <DefaultValue(ComboBoxStyle.DropDown)> _
    Public Property DropDownStyle() As ComboBoxStyle
        Get
            Return ComboBox.DropDownStyle
        End Get

        Set(ByVal value As ComboBoxStyle)
            ComboBox.DropDownStyle = value
        End Set
    End Property

    <Category("Appearance"), Description("Occurs when a visual aspect of an owner-drawn ComboBox changes.")> _
    Public Event DrawItem As System.Windows.Forms.DrawItemEventHandler
    ''' <summary>
    ''' Raises the <see cref="DrawItem"/> event.
    ''' </summary>
    Protected Overridable Sub OnDrawItem(ByVal e As System.Windows.Forms.DrawItemEventArgs)
        RaiseEvent DrawItem(ComboBox, e)
    End Sub

    <Category("Property Changed")> _
    Public Event SelectedValueChanged As EventHandler
    ''' <summary>
    ''' Raises the <see cref="SelectedValueChanged"/> event.
    ''' </summary>
    Protected Overridable Sub OnSelectedValueChanged(ByVal e As EventArgs)
        RaiseEvent SelectedValueChanged(ComboBox, e)
    End Sub

#End Region

#Region "** methods"

    Protected Overrides Sub OnSubscribeControlEvents(ByVal ctrl As Control)
        MyBase.OnSubscribeControlEvents(ctrl)
        AddHandler CType(ctrl, ComboBox).DrawItem, AddressOf HandleDrawItem
        AddHandler CType(ctrl, ComboBox).SelectedValueChanged, AddressOf HandleSelectedValueChanged
    End Sub

    Protected Overrides Sub OnUnsubscribeControlEvents(ByVal ctrl As Control)
        MyBase.OnUnsubscribeControlEvents(ctrl)
        RemoveHandler CType(ctrl, ComboBox).DrawItem, AddressOf HandleDrawItem
        RemoveHandler CType(ctrl, ComboBox).SelectedValueChanged, AddressOf HandleSelectedValueChanged
    End Sub

    Private Sub HandleDrawItem(ByVal sender As Object, ByVal e As System.Windows.Forms.DrawItemEventArgs)
        Me.OnDrawItem(e)
    End Sub

    Private Sub HandleSelectedValueChanged(ByVal sender As Object, ByVal e As EventArgs)
        Me.OnSelectedValueChanged(e)
    End Sub

#End Region

End Class

File: NetFramework/Schedule/CS/MultiUserCustomForms/RibbonControls/ComboBoxHost.cs
using System.Windows.Forms;
using C1.Win.C1Ribbon;
using System.ComponentModel;
using System;
using System.Drawing.Design;

namespace RibbonControls
{
	/// <summary>
	/// The <see cref="ComboBoxHost"/> control wraps standard ComboBox control 
	/// so that it can be used in the <see cref="C1Ribbon"/> control.
	/// </summary>
	[System.ComponentModel.DefaultBindingProperty("SelectedValue")]
	public class ComboBoxHost : RibbonControlHost, IBindableComponent
	{
		#region ** ctor
		public ComboBoxHost()
			: this(new ComboBox())
		{
		}
		public ComboBoxHost(ComboBox box) 
			: base (box)
		{
		}
		#endregion

		#region ** ComboBox & ComboBox interface wrappers
		[Browsable(false)]
		[DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
		public ComboBox ComboBox
		{
			get
			{
				return (ComboBox)base.Control;
			}
		}

		[Category("Data"), Browsable(false), Bindable(true),
		DefaultValue(null),
		DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
		public object SelectedValue
		{
			get
			{
				return ComboBox.SelectedValue;
			}
			set
			{
				ComboBox.SelectedValue = value;
			}
		}

		[Category("Data"), Browsable(false), Bindable(true),
		DefaultValue(null),
		DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
		public object SelectedItem
		{
			get
			{
				return ComboBox.SelectedItem;
			}
			set
			{
				ComboBox.SelectedItem = value;
			}
		}

		[Category("Data"), Description("Indicates the list that this control will use to get its items."),
		AttributeProvider("System.ComponentModel.IListSource"),
		DefaultValue(null),
		RefreshProperties(RefreshProperties.Repaint)]
		public object DataSource
		{
			get
			{
				return ComboBox.DataSource;
			}
			set
			{
				ComboBox.DataSource = value;
			}
		}
		[Category("Data"), Description("Indicates the property to display for the items in the control."),
		Editor("System.Windows.Forms.Design.DataMemberFieldEditor, System.Design", "System.Drawing.Design.UITypeEditor"),
		TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, System.Design"),
		DefaultValue("")]
		public string DisplayMember
		{
			get
			{
				return ComboBox.DisplayMember;
			}

			set
			{
				ComboBox.DisplayMember = value;
			}
		}

		[Category("Data"), Description("Indicates the property to use as the actual value for the items in the control."),
		Editor("System.Windows.Forms.Design.DataMemberFieldEditor, System.Design", "System.Drawing.Design.UITypeEditor"),
		TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, System.Design"),
		DefaultValue("")]
		public string ValueMember
		{
			get
			{
				return ComboBox.ValueMember;
			}
			set
			{
				ComboBox.ValueMember = value;
			}
		}
		[Category("Data"), Description("Gets the collection of items in this ComboBox.")]
		[DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
		[Editor("System.Windows.Forms.Design.ListControlStringCollectionEditor, System.Design", typeof(UITypeEditor))]
		public ComboBox.ObjectCollection Items
		{
			get
			{
				return ComboBox.Items;
			}
		}

		[Category("Behavior"), Description("Indicates whether the code or the operating system will handle drawing of elements in the list.")]
		[DefaultValue(DrawMode.Normal)]
		public DrawMode DrawMode
		{
			get
			{
				return ComboBox.DrawMode;
			}
			set
			{
				ComboBox.DrawMode = value;
			}
		}

		[Category("Appearance"), Description("Controls the appearance and functionality of the ComboBox.")]
		[DefaultValue(ComboBoxStyle.DropDown)]
		public ComboBoxStyle DropDownStyle
		{
			get
			{
				return ComboBox.DropDownStyle;
			}
			set
			{
				ComboBox.DropDownStyle = value;
			}
		}

		/// <summary>
		/// Raises the <see cref="DrawItem"/> event.
		/// </summary>
		[Category("Appearance"), Description("Occurs when a visual aspect of an owner-drawn ComboBox changes.")]
		public event System.Windows.Forms.DrawItemEventHandler DrawItem; 
		protected virtual void OnDrawItem(System.Windows.Forms.DrawItemEventArgs e)
		{
			if (DrawItem != null)
			{
				DrawItem(ComboBox, e);
			}
		}
		[Category("Property Changed")]
		public event EventHandler SelectedValueChanged;
		protected virtual void OnSelectedValueChanged(EventArgs e)
		{
			if (SelectedValueChanged != null)
			{
				SelectedValueChanged(ComboBox, e);
			}
		}

		protected override void OnSubscribeControlEvents(Control control)
		{
			base.OnSubscribeControlEvents(control);
			ComboBox.DrawItem += new System.Windows.Forms.DrawItemEventHandler(ComboBox_DrawItem);
			ComboBox.SelectedValueChanged += new EventHandler(ComboBox_SelectedValueChanged);
		}

		protected override void OnUnsubscribeControlEvents(Control control)
		{
			base.OnUnsubscribeControlEvents(control);
			ComboBox.DrawItem -= new System.Windows.Forms.DrawItemEventHandler(ComboBox_DrawItem);
			ComboBox.SelectedValueChanged -= new EventHandler(ComboBox_SelectedValueChanged);
		}

		void ComboBox_DrawItem(object sender, System.Windows.Forms.DrawItemEventArgs e)
		{
			OnDrawItem(e);
		}
		void ComboBox_SelectedValueChanged(object sender, EventArgs e)
		{
			OnSelectedValueChanged(e);
		}

		#endregion

		#region ** IBindableComponent Members
		[Browsable(false)]
		[DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
		public BindingContext BindingContext
		{
			get
			{
				return ComboBox.BindingContext;
			}
			set
			{
				ComboBox.BindingContext = value;
			}
		}

		[Category("Data")]
		public ControlBindingsCollection DataBindings
		{
			get { return ComboBox.DataBindings; }
		}
		#endregion


	}
}